### ISSUE TITLE
Store OSB client metadata as CF service instance annotations (missing upstream PR + update support)

### ISSUE THREAD

**Issue by gberche-orange** [orange-cloudfoundry/osb-cmdb-spike#10]
### Expected behavior

As an osb-cmdb user, in order to fetch the cmdb by the OSB client meta-data (e.g. orange corporate application code "basicat", CF org/space/originating user/annotations ..., K8S namespace/annotations ), I need the OSB client medata to be stored as CF service instance annotations

### Observed behavior

Osb cmdb is currently ignoring metadata provided by the OSB client.

### Affected release

Reproduced on version 0.2.0

**Comment by gberche-orange** [comment]
Status as of paas-templates v45 (i.e. cf-deployment  12.10.0 ):

### only set labels/annotations on a service instance with cf curl, no read endpoint available

``` 
 CF_TRACE=true cf curl v3/service_instances/$(cf service p-mysql-0d5b8135-2953-4109-99f5-7bb0b048ec4f --guid)/ -X PATCH -d '{ "metadata": {"labels": {"LABEL-KEY": "LABEL-VALUE" } } }'
 CF_TRACE=true cf curl v3/service_instances/$(cf service p-mysql-0d5b8135-2953-4109-99f5-7bb0b048ec4f --guid)/ -X PATCH -d '{ "metadata": {"annotations": {"ANNOTATION-KEY": "ANNOTATION-VALUE" } } }'

cf curl v3/service_instances/$(cf service p-mysql-0d5b8135-2953-4109-99f5-7bb0b048ec4f --guid)
{
   "errors": [
      {
         "detail": "Unknown request",
         "title": "CF-NotFound",
         "code": 10000
      }
   ]
}

``` 
   - No get Read endpoint for services http://v3-apidocs.cloudfoundry.org/version/3.79.0/index.html#service-instances
   - The SAPI backlog [v3 epic](https://www.pivotaltracker.com/epic/show/4129144) does not yet lists stories for service instances
   - The SAPI PM [indicated roadmap](https://docs.google.com/document/d/1fVO3nA0T5uApN5yWqSZaD3ddZQClCzogeivKxhriLJs/edit?disco=AAAAD4ycn-s) as 

> We're doing things in this order: service brokers, offerings, plans, plan-visibilities, .. and then instances. I've put together a visual view on our Aha: https://pcf.aha.io/products/ENABLEMNT/feature_cards .. however please take the timeline as verrry indicative - as we're still trying to figure out our progress velocity etc.
   - The CLI backlog [meta-data epic](https://www.pivotaltracker.com/epic/show/4231034) 
   - The VAT (V3 acceleration team) backlog [meta-data label](https://www.pivotaltracker.com/n/projects/2196383/search?q=label%3A%22metadata%22) only lists domain, route, and users
   - The SAPI backlog currently only includes [metadata stories](https://www.pivotaltracker.com/n/projects/2105761/search?q=label%3A%22metadata%22%20includedone%3Atrue) for the service-broker v3 resource


### labels on spaces are working as expected in CLI

- labels on spaces are working as expected, but contain [documented character set restrictions](https://docs.cloudfoundry.org/adminguide/metadata.html#reqs): only alpha numeric characters, 253 for key, 63 for values

```
$ cf7 labels space p-mysql
Getting labels for space p-mysql in org osb-cmdb-backend-services-org-client-1 as xx...

key                        value
KEY                        VALUE
service-guid1-org-guid     guid2
service-guid1-space-guid   guid1
service-guid1-user-guid    guid3
```

### annotations on spaces are working as expected with cf curl 

```
CF_TRACE=true cf curl /v3/spaces/2de0ae7a-ace7-412d-9aac-9d36ce3fbed9 -X PATCH -d '{ "metadata": {"annotations": {"service-guid1-organization_name": "a long name with spaces and arbitrary characters" } } }'
CF_TRACE=true cf curl /v3/spaces/2de0ae7a-ace7-412d-9aac-9d36ce3fbed9 -X PATCH -d '{ "metadata": {"annotations": {"service-guid1-space_name": "a long space name with spaces and arbitrary characters" } } }'
CF_TRACE=true cf curl /v3/spaces/2de0ae7a-ace7-412d-9aac-9d36ce3fbed9 -X PATCH -d '{ "metadata": {"annotations": {"service-guid1-organization_name": "a long org name with spaces and arbitrary characters" } } }'
CF_TRACE=true cf curl /v3/spaces/2de0ae7a-ace7-412d-9aac-9d36ce3fbed9 -X PATCH -d '{ "metadata": {"annotations": {"service-guid1-instance_name": "a long instance name with spaces and arbitrary characters" } } }'

cf curl /v3/spaces/2de0ae7a-ace7-412d-9aac-9d36ce3fbed9 | jq .metadata
{
  "labels": {
    "service-guid1-user-guid": "guid5",
    "LABEL-KEY": "LABEL-VALUE",
    "service-guid1-org-guid": "guid2",
    "service-guid1-space-guid": "guid1",
    "KEY": "VALUE"
  },
  "annotations": {
    "service-guid1-instance_name": "a long instance name with spaces and arbitrary characters",
    "service-guid1-organization_name": "a long org name with spaces and arbitrary characters",
    "service-guid1-space_name": "a long space name with spaces and arbitrary characters"
  }
}
```


- CF CLI annotations is not yet supported, see related [story template](https://www.pivotaltracker.com/epic/show/4466010) and [metadata epic: annotations not yet started](https://www.pivotaltracker.com/epic/show/4231034)

**Comment by gberche-orange** [comment]
### Use case 1: lookup backing service from osb customer data

- As a service operator
- In order to respond to incoming support request from an OSB customer
- I need to find a backing service instance from any combination of osb client name, org id,  space id, service instance id, user id
- And I need to find a backing service instance from any combination of osb client name, org name, space name, service instance name (longer 


### Use case 2: lookup osb customer meta-data from backing service guid

- As a service operator
- In order to identify the OSB customer associated to a backing service instance
- I need to lookup the osb client name, org name, space name, service instance name, user id, service guid for a given backing service instance guid

### Labels

Label name | Label value | Query usage | Read usage
-- | -- | -- | --
backing-service-instance-guid | backing-service-instance-guid | workaround lack of service instance read endpoint |  | 
brokered-service-guid | brokered-service-guid | find a backend service by brokered-service-guid in all orgs/spaces | (redundant with backend service instance name)  | 
brokered-service-meta-org-guid | brokered-service-meta-org-guid | find all backing services for a brokered-service-org-guid | lookup org guid for a given backing service | 
brokered-service-meta-space-guid | brokered-service-meta-space-guid | find all backing services for a brokered-service-space-guid | lookup space guid for a given backing service | 
brokered-service-meta-user-guid | brokered-service-meta-user-guid | find all backing services for a brokered-service-user-guid | lookup user guid for a given backing service | 
~~brokered-service-meta-osb-client-id~~ | brokered-service-meta-user-guid | find all backing services for a osb client id **(redundant with org)** | lookup osb client id for a given backing service **(redundant with org)**| 

### Annotations

Annotation name | Annotation value | ~~Query usage~~ | Read usage
-- | -- | -- | --
brokered-service-meta-org-name |  | | lookup org name for a given backing service | 
brokered-service-meta-space-name |  | |  lookup space name for a given backing service| 
brokered-service-meta-service-instance-name | |  | | lookup service instance name for a given backing service |

**Comment by gberche-orange** [comment]
Cf java client supports annotations, with ListUSI and UpdateSI with labels and annotations:
- https://github.com/cloudfoundry/cf-java-client/blob/1033a9dc11/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/_Metadata.java
- https://github.com/cloudfoundry/cf-java-client/blob/1033a9dc11b2d28ec26577ca39bb02392a035c77/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/_UpdateServiceInstanceRequest.java
- https://github.com/cloudfoundry/cf-java-client/blob/cb4f2b95513816d37b45080e1810c43fb63964f5/cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/serviceinstances/ReactorServiceInstancesV3.java#L74-L77
- https://github.com/cloudfoundry/cf-java-client/blob/cb4f2b95513816d37b45080e1810c43fb63964f5/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/ServiceInstancesV3.java#L68
- https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/_ListServiceInstancesRequest.java
   - https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/_ListServiceInstancesResponse.java
   - https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/_ServiceInstanceResource.java
   - https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/serviceInstances/ServiceInstance.java


Implementations steps:
* attempt to create acceptance test (discover cf-java-client api)
* attempt to create component test ?
* implement unit tests for CF deployer

**Comment by gberche-orange** [comment]
Fixed in v1.0.0 See https://github.com/orange-cloudfoundry/osb-cmdb-spike/blob/67f7cdd3cd1e707ede5c7f31d015053a4f86a268/osb-cmdb/src/main/java/com/orange/oss/osbcmdb/serviceinstance/OsbCmdbServiceInstance.java#L410-L414

Acceptance test is not asserting this yet https://github.com/orange-cloudfoundry/osb-cmdb-spike/blob/v1.0.0/spring-cloud-app-broker-acceptance-tests/src/test/java/org/springframework/cloud/appbroker/acceptance/UpdateInstanceWithBackingServiceAcceptanceTest.java
